### PR TITLE
refactor(turborepo): Signature Authentication (2nd try)

### PR DESCRIPTION
**Pull request by NicholasLYang**
### Description

Refactors the signature authentication to use a language-independent algorithm, as the previous implementation depended on Go's JSON serialization.

NOTE: This is a global hash bump.

### Testing Instructions

Testing is using fuzzing across the Go-Rust boundary to check that signatures are correctly generated on both sides.

### CODE REVIEW

File: crates/turborepo-cache/src/signature_authentication.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+impl ArtifactSignatureAuthenticator {
+    pub fn new(team_id: Vec<u8>, secret_key_override: Option<Vec<u8>>) -> Self {
+        Self {
+            team_id,
+            secret_key_override,
+        }
+    }
+
+    fn secret_key(&self) -> Result<Vec<u8>, SignatureError> {
```
Comment (arlyon) [inline crates/turborepo-cache/src/signature_authentication.rs:39]: I believe that HMAC_SHA256 doesn't have key length requirements but a comment that verifies that would be handy.

File: crates/turborepo-cache/src/signature_authentication.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        Ok(BASE64_STANDARD.encode(hmac_output))
+    }
+
+    pub fn validate(
+        &self,
+        hash: &[u8],
+        artifact_body: &[u8],
+        expected_tag: &str,
+    ) -> Result<bool, SignatureError> {
+        let secret_key = hmac::Key::new(HMAC_SHA256, &self.secret_key()?);
```
Comment (arlyon) [inline crates/turborepo-cache/src/signature_authentication.rs:96]: I also think we could pull HMAC_SHA256 into a constant

**Review by arlyon — APPROVED** [review]
Approved with some minor comments. I think we should announce this in the turbo channel / turboverse before merging.

File: crates/turborepo-cache/src/signature_authentication.rs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+         TURBO_REMOTE_CACHE_SIGNATURE_KEY environment variable"
+    )]
+    NoSignatureSecretKey,
+    #[error("serialization error: {0}")]
+    SerializationError(#[from] serde_json::Error),
+    #[error("base64 encoding error: {0}")]
+    Base64EncodingError(#[from] base64::DecodeError),
+}
+
+static TURBO_HMAC_ALGORITHM: Algorithm = HMAC_SHA256;
```
Comment (NicholasLYang) [inline crates/turborepo-cache/src/signature_authentication.rs:24]: this is static because apparently you can't have a const that refers to another const

**Pull request merged**